polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -90,36 +84,249 @@ static bool isSmallerThanNegativeOne(Value val, DataFlowSolver &solver) {
+operator<<(raw_ostream &os, const ValueOrMultiplier &valOrMultiplier) {
+  if (valOrMultiplier.isMultiplier())
+    return os << valOrMultiplier.getMultiplier();
+  return os << valOrMultiplier.getValue();
+}
+
+} // namespace
+
+// Visit a binary operation of type \tparam T. The LHS and RHS operand of the
+// binary operation are processed by applying the function \p getMultiplier. The
+// result of the visit is computed via the \p computeResult function.
+template <typename T, typename ProcessOperandFuncT, typename ComputeResultFuncT,
+          typename = std::enable_if_t<llvm::is_one_of<
+              T, arith::AddIOp, arith::SubIOp, arith::MulIOp>::value>,
+          typename = std::enable_if<
+              std::is_invocable_r_v<ValueOrMultiplier, ProcessOperandFuncT, T,
+                                    const Value, DataFlowSolver &>>,
+          typename = std::enable_if<
+              std::is_invocable_r_v<ValueOrMultiplier, ComputeResultFuncT,
+                                    ValueOrMultiplier, ValueOrMultiplier>>>
+static ValueOrMultiplier visitBinaryOp(T binOp, const Value factor,
+                                       DataFlowSolver &solver,
+                                       ProcessOperandFuncT getMultiplier,
+                                       ComputeResultFuncT computeResult) {
+  // Traverse my subtrees.
+  auto lhsRes = getMultiplier(binOp.getLhs(), factor, solver);
+  auto rhsRes = getMultiplier(binOp.getRhs(), factor, solver);
+
+  // Compute the result to propagate up.
+  auto res = computeResult(lhsRes, rhsRes);
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "lhsRes: " << lhsRes << "\n";
+    llvm::dbgs() << "rhsRes: " << rhsRes << "\n";
+    llvm::dbgs() << "res = " << res << "\n";
+  });
+
+  return res;
+}
+
+/// Determine whether \p expr involves a subexpression which is a multiplication
+/// of \p val, and return the multiplication factor if it does.
+/// Example:
+///   %c1_i32 = arith.constant 1 : i32
+///   %c2_i32 = arith.constant 2 : i32
+///   %index_cast = arith.index_cast %ii : index to i64
+///   %c1_i64 = arith.extsi %c1_i32 : i32 to i64
+///   %mul = arith.muli %index_cast, %c1_i64 : i64
+///   %c2_i64 = arith.extsi %c2_i32 : i32 to i64
+///   %add = arith.addi %mul, %c2_i64 : i64
+///
+/// Here getMultiplier(%add, %ii) should return '%c1_i32'.
+static ValueOrMultiplier getMultiplier(const Value expr, const Value factor,
+                                       DataFlowSolver &solver) {
+  Operation *op = expr.getDefiningOp();
+  if (!op) {
+    // This is a block argument. If it is a match create a multiplier with value
+    // one.
+    if (expr == factor)
+      return Multiplier::one(expr.getContext());
+    return expr;
+  }
+
+  auto getOperandThatMatchesFactor = [&factor](Value lhs, Value rhs) {
+    if (lhs == factor && lhs != rhs)
+      return lhs;
+    if (rhs == factor && lhs != rhs)
+      return rhs;
+    return Value();
+  };
+
+  return TypeSwitch<Operation *, ValueOrMultiplier>(op)
```

Comment:
I'm a bit worried we end up doing too much "expression analysis" here. As we only care about integer arithmetic here, we could actually have a `-reassociate` pass ([as in LLVM](https://llvm.org/docs/Passes.html#reassociate-reassociate-expressions)) on `arith` operations, which should greatly improve the cases we should detect.
I think that would add way more value compared to this ad-hoc analysis as it could also potentially improve performance, WDYT? Also, this could be upstreamed.

Reply:
After the offline discussion, the current approach LGTM. We can move quickly as is and we can think of the `-reassociate` pass in the future.